### PR TITLE
Customer/address

### PR DESCRIPTION
**Pull request by julienbourdeau**
I introduce the address form.

I would like feedback about the error management and especially

``` smarty
{if $data.required}
  {block name="required_field"}{include file="_partials/form-required-field.tpl"}{/block}
{/if}
```

and 

``` smarty
{block name="form_field_error"}{include file="_partials/form-field-errors.tpl" errors=$form_errors[$field_name]}{/block}
```

The form with error looks like
address_-_prestadev_1_7

and the HTML markup:

``` html
<label>
    <span>Address</span>
    <sup>*</sup>
    <input type="text" name="address1" id="address1" value="">
    <p>This information is required.</p>
</label>
```

### CODE REVIEW

**Comment by dkarvounaris** [comment]
My suggestions: 

Each label should have a "for" attribute, this is pretty standard and activates several useful behaviours for the browser and/or jquery plugins. I know this is usually implied when the `<input>` is inside the `<label>`, but should make modifications faster and less error-prone by not having to add them if someone needs to move it outside. Also makes it easier to be referenced through jquery, javascript or css this way. 

The <input> field could be something else in some cases... there should be a way to address this without knowing what it is. One such way could be:

``` html
<fieldset class>
    <label for="field-id">Address <sup>*</sup></label>
    <div class><input type="text" name="address1" id="field-id" value=""></div>
    <p>This information is required.</p>
</fieldset>
```

This example would even allow for more css stylings, also creative ones (as example label showing inside div, until selected, then minimizing it above, all within the same visual border created by the fieldset) without template changes.

**Comment by dkarvounaris** [comment]
Also think about when having more than one `<input>` or `<select>` assigned to a label (example: birthdate). Having them in a `<div>` is useful then for formatting and I've found having the "for" attribute helps some browsers who act unpredicted otherwise in this case.

**Comment by gskema** [comment]
Fieldset should contain multiple inputs inside. Therefore I'd be better to wrap the input row in a `<div>`. Label should be definitely separated from input (lets think of accessibility).

As for the errors, I think that `form-required-field.tpl` won't prove to be that useful:
- I may want to put a `required` class or some other attribute on the input wrapper, eg. `.form-group required`
- Why put `<sup>*</sup>` outside, why not put the whole `label` in its own partial?
- I may want to put `required=""` attribute on the input (which we do in practive a lot)

Also, I want to point out that `$form_errors` should be formed in such way that we are able to:
- list all errors at once above the form
- add classes (`form-error`, `form-success`) to input wrapper
- output one or more errors below each specific field

**Comment by julienbourdeau** [comment]
#### About `for` attribute

I was thinking that in this case you need the field id to be unique and I wasn't sure, like if you include the for login form in your header and in the page content. Or if you use fancybox or ajax. Maybe this is just an edge case and I shouldn't worry about it.

#### About errors

You can display many errors, it's an array of strings. I also removed `$form_errors`, errors are in the `$ordered_address_fields` variable so it's very easy to use and add classes if needed (inside the foreach). I'll think about something to add classes outside the foreach.

#### About `<fieldset>`

As far as I understand, fieldset should be used for a group of fields, like if we group b2b related fields, personal info related fields,... I don't think we should use it in the starter theme. I feel like it's pretty old as well, but maybe it's just me.

#### About `required` attribute

I like using a partial for `<sup>*</sup>` so you can easy change it. Like if you, as a theme developer, you want to use `<strong>{l s='Required'}</strong>` it's easy to change and all forms use the same.
The required attribute is already handled.

``` smarty
{if $data.required}required{/if}
```

#### About a label block partials

That would be amazing :D I only need to sort out the translation issue !

**Comment by julienbourdeau** [comment]
The form now looks like :

```
Array
(
    [postcode] => Array
        (
            [required] => 
            [errors] => Array
                (
                    [0] => The Zip/Postal code you&#039;ve entered is invalid. It must follow this format: 000000
                )

            [label] => Zip/Postal Code
        )

    [Country:name] => Array
        (
            [required] => 1
            [errors] => Array
                (
                )

            [label] => Country
        )

    [phone] => Array
        (
            [required] => 
            [errors] => Array
                (
                    [0] => You must register at least one phone number.
                )

            [label] => Home phone
        )

    [phone_mobile] => Array
        (
            [required] => 
            [errors] => Array
                (
                    [0] => You must register at least one phone number.
                )

            [label] => Mobile phone
        )

)
```

So the template is simpler.

**Comment by gskema** [comment]
- I think `item=data}` should be renamed to `item=field}`
- Also, if we are using so many blocks, why aren't we just using input-type based partials like (someone enlighten me, I'm just throwing and idea here):

```
{include file="_partials/form-field-text.tpl" field=$field}
...
{include file="_partials/form-field-select.tpl" field=$field}
```

If you need a really specific input widget you can create a new template. Don't know if these would be for address only, or for many forms.'

The label part is interesting too. Where can we edit them?
